### ISSUE TITLE
Calling entry.SetPlaceHolder inside of OnChanged callback freezes app

### ISSUE THREAD

**Issue by dweymouth** [fyne-io/fyne#4516]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

Reported in Fyne discord.

Calling entry.SetPlaceHolder in the OnChanged callback when it's invoked from a select-all + backspace change (ie deleting all text) causes the app to freeze. It reportedly works in Fyne 2.3.5

### How to reproduce

Run the example code. Type into the entry. Then Select all (Ctrl+A) and delete.

### Example code

```go
package main

import (
    "fyne.io/fyne/v2/app"
    "fyne.io/fyne/v2/widget"
)

func main() {
    myApp := app.New()
    window := myApp.NewWindow("test")

    str := "hello"
    entry := widget.NewEntry()
    entry.SetText(str)
    entry.OnChanged = func(s string) {
        if entry.Text == "" {
            entry.SetPlaceHolder("bye")
        } else {
            str = s
        }
    }
    window.SetContent(entry)
    window.ShowAndRun()
}
```

### Fyne version

2.4.3

### Go compiler version

1.21.4

### Operating system and version

Windows

**Comment by Jacalz** [comment]
I think this should be a quite simple fix. We just have to look at all the places we are calling `OnChanged` and make sure we aren't holding any locks when doing so. A quick glance shows this:

https://github.com/fyne-io/fyne/blob/a09bbd790b98120bc406a064f80aeafb4ca97656/widget/entry.go#L1097

https://github.com/fyne-io/fyne/blob/a09bbd790b98120bc406a064f80aeafb4ca97656/widget/entry.go#L851

**Comment by Jacalz** [comment]
Fixed in upcoming v2.4.4 release
